quil-rs/src/program/scheduling/schedule.rs

```diff
@@ -0,0 +1,479 @@
+//! A Schedule represents a flattening of the [`DependencyGraph`] into a linear sequence of
+//! instructions, with each instruction assigned a start time and duration.
+
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use petgraph::{
+    visit::{EdgeFiltered, Topo},
+    Direction,
+};
+
+use crate::{
+    instruction::{
+        AttributeValue, Capture, Delay, Instruction, Pulse, RawCapture, WaveformInvocation,
+    },
+    quil::Quil,
+    Program,
+};
+
+use super::{ExecutionDependency, ScheduledBasicBlock, ScheduledGraphNode};
+
+#[derive(Clone, Debug, Default, PartialEq, PartialOrd)]
+pub struct Seconds(pub f64);
+
+impl std::ops::Add<Seconds> for Seconds {
+    type Output = Seconds;
+
+    fn add(self, rhs: Seconds) -> Self::Output {
+        Self(rhs.0 + self.0)
+    }
+}
+
+impl std::ops::Sub<Seconds> for Seconds {
+    type Output = Seconds;
+
+    fn sub(self, rhs: Seconds) -> Self::Output {
+        Self(self.0 - rhs.0)
+    }
+}
+
+pub trait Zero: PartialEq + Sized {
+    fn zero() -> Self;
+
+    fn is_zero(&self) -> bool {
+        self == &Self::zero()
+    }
+}
+
+impl Zero for Seconds {
+    fn zero() -> Self {
+        Self(0.0)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Schedule<Time> {
+    items: Vec<ComputedScheduleItem<Time>>,
+    /// The total duration of the block. This is the end time of the schedule when it starts at `Time::zero()`
+    duration: Time,
+}
+
+impl<Time> Schedule<Time> {
+    pub fn duration(&self) -> &Time {
+        &self.duration
+    }
+
+    pub fn items(&self) -> &[ComputedScheduleItem<Time>] {
+        self.items.as_ref()
+    }
+
+    pub fn into_items(self) -> Vec<ComputedScheduleItem<Time>> {
+        self.items
+    }
+}
+
+impl<Time: Clone + PartialOrd + std::ops::Add<Time, Output = Time> + Zero>
+    From<Vec<ComputedScheduleItem<Time>>> for Schedule<Time>
+{
+    fn from(items: Vec<ComputedScheduleItem<Time>>) -> Self {
+        let duration = items
+            .iter()
+            .map(|item| item.time_span.start_time.clone() + item.time_span.duration.clone())
+            .fold(Time::zero(), |acc, el| if el > acc { el } else { acc });
```

Comment:
Same question as on the Python doc-comment for `duration`: aren't the "items" here instructions within a block, and doesn't this need to account for possible dependencies between them?

Reply:
The dependencies are handled in the scheduling of `ScheduledBasicBlock.as_fixed_schedule`.

If instruction `GATE 0` expands into 

```
PULSE
DELAY
PULSE
```

Then this just unions/covers the times of the PULSE, DELAY, PULSE and considers that combined TimeSpan as the TimeSpan of the original gate